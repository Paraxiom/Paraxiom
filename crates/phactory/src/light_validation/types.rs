pub use phactory_api::blocks::{AuthoritySet, AuthoritySetChange};
