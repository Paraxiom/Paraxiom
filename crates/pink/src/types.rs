use sp_runtime::{AccountId32, traits::BlakeTwo256};

pub use frame_support::weights::Weight;

pub type Hash = sp_core::H256;
pub type Hashing = BlakeTwo256;
pub type AccountId = AccountId32;
pub type Balance = u128;
pub type BlockNumber = u32;
pub type Index = u64;
