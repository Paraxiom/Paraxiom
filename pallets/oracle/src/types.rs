use core::default;

use crate::Config;
use frame_support::pallet_prelude::{Decode, Encode, MaxEncodedLen, TypeInfo};
use frame_support::{
    traits::{ConstU128, ConstU32, ConstU64},
    BoundedVec
};
use sp_core::H256;


pub type RequestId = H256;

#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
#[scale_info(skip_type_params(T))]
pub struct Request<T: Config> { 
    pub requester: T::AccountId,
    pub nonce: u64,
}