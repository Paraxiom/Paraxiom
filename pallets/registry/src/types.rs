use frame_support::pallet_prelude::{Decode, Encode, MaxEncodedLen, TypeInfo};
use frame_support::{
    traits::{ConstU128, ConstU32, ConstU64},
    BoundedVec,
};
use crate::Config;

/// Types representing limited strings
pub type RegistryFeedKey<T> = BoundedVec<u8, <T as Config>::MaxKeySize>;
pub type RegistryFeedUrl<T> = BoundedVec<u8, <T as Config>::MaxUrlSize>;
pub type RegistryFeedPath<T> = BoundedVec<u8, <T as Config>::MaxPathSize>;

/// Feed status
#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
#[cfg_attr(feature = "std", derive(Debug))]
pub enum ApiFeedStatus {
    Registered,
    Active,
    Inactive,
    // occurs if a feed has too many errors
    Slashed,
}

impl Default for ApiFeedStatus {
    fn default() -> Self {
        ApiFeedStatus::Registered
    }
}