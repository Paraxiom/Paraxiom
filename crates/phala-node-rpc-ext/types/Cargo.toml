[package]
name = "phala-node-rpc-ext-types"
version = "0.1.0"
edition = "2021"

[dependencies]
serde = { version = "1.0", features = ["derive"] }
impl-serde = "0.4.0"
scale-info = { version = "2.1", default-features = false }
scale = { package = "parity-scale-codec", version = "3.1" }
