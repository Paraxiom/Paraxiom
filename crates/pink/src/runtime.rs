mod extension;
mod pallet_pink;
mod weights;

use std::{
    sync::RwLock,
    time::{Duration, Instant},
};

use crate::types::{AccountId, Balance, BlockNumber, Hash, Hashing, Index};
use frame_support::{
    parameter_types,
    weights::{constants::WEIGHT_PER_SECOND, Weight},
};
use pallet_contracts::{Config, Frame, Schedule};
use sp_runtime::{generic::Header, traits::IdentityLookup, Perbill};

pub use extension::{get_side_effects, ExecSideEffects};
pub use pink_extension::{EcdhPublicKey, HookPoint, Message, OspMessage, PinkEvent};

type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<PinkRuntime>;
type Block = frame_system::mocking::MockBlock<PinkRuntime>;

frame_support::construct_runtime! {
    pub enum PinkRuntime where
        Block = Block,
        NodeBlock = Block,
        UncheckedExtrinsic = UncheckedExtrinsic,
    {
        System: frame_system,
        Timestamp: pallet_timestamp,
        Balances: pallet_balances,
        Randomness: pallet_randomness_collective_flip,
        Contracts: pallet_contracts,
        Pink: pallet_pink,
    }
}

const NORMAL_DISPATCH_RATIO: Perbill = Perbill::from_percent(75);

parameter_types! {
    pub const BlockHashCount: u32 = 250;
    pub RuntimeBlockWeights: frame_system::limits::BlockWeights =
        frame_system::limits::BlockWeights::with_sensible_defaults(
            (2u64 * WEIGHT_PER_SECOND).set_proof_size(u64::MAX),
            NORMAL_DISPATCH_RATIO,
        );
    pub const ExistentialDeposit: Balance = 1;
    pub const MaxLocks: u32 = 50;
    pub const MaxReserves: u32 = 50;
}

impl pallet_pink::Config for PinkRuntime {
    type Currency = Balances;
}

impl pallet_balances::Config for PinkRuntime {
    type Balance = Balance;
    type DustRemoval = ();
    type RuntimeEvent = RuntimeEvent;
    type ExistentialDeposit = ExistentialDeposit;
    type AccountStore = frame_system::Pallet<PinkRuntime>;
    type WeightInfo = pallet_balances::weights::SubstrateWeight<PinkRuntime>;
    type MaxLocks = MaxLocks;
    type MaxReserves = MaxReserves;
    type ReserveIdentifier = [u8; 8];
}

impl frame_system::Config for PinkRuntime {
    type BaseCallFilter = frame_support::traits::Everything;
    type BlockWeights = RuntimeBlockWeights;
    type BlockLength = ();
    type DbWeight = ();
    type RuntimeOrigin = RuntimeOrigin;
    type Index = Index;
    type BlockNumber = BlockNumber;
    type Hash = Hash;
    type RuntimeCall = RuntimeCall;
    type Hashing = Hashing;
    type AccountId = AccountId;
    type Lookup = IdentityLookup<Self::AccountId>;
    type Header = Header<Self::BlockNumber, Self::Hashing>;
    type RuntimeEvent = RuntimeEvent;
    type BlockHashCount = BlockHashCount;
    type Version = ();
    type PalletInfo = PalletInfo;
    type AccountData = pallet_balances::AccountData<Balance>;
    type OnNewAccount = ();
    type OnKilledAccount = ();
    type SystemWeightInfo = ();
    type SS58Prefix = ();
    type OnSetCode = ();
    type MaxConsumers = frame_support::traits::ConstU32<16>;
}

impl pallet_randomness_collective_flip::Config for PinkRuntime {}

parameter_types! {
    pub const MinimumPeriod: u64 = 1;
}

impl pallet_timestamp::Config for PinkRuntime {
    type Moment = u64;
    type OnTimestampSet = ();
    type MinimumPeriod = MinimumPeriod;
    type WeightInfo = ();
}

parameter_types! {
    pub DepositPerStorageByte: Balance = Pink::deposit_per_byte();
    pub DepositPerStorageItem: Balance = Pink::deposit_per_item();
    pub const DeletionQueueDepth: u32 = 1024;
    pub const DeletionWeightLimit: Weight = Weight::from_ref_time(500_000_000_000);
    pub const MaxCodeLen: u32 = 2 * 1024 * 1024;
    pub const MaxStorageKeyLen: u32 = 128;

    pub DefaultSchedule: Schedule<PinkRuntime> = {
        let mut schedule = Schedule::<PinkRuntime>::default();
        const MB: u32 = 16;  // 64KiB * 16
        // Each concurrent query would create a VM instance to serve it. We couldn't
        // allocate too much here.
        schedule.limits.memory_pages = 4 * MB;
        schedule.instruction_weights.fallback = 8000;
        schedule
    };
}

impl Config for PinkRuntime {
    type Time = Timestamp;
    type Randomness = Randomness;
    type Currency = Balances;
    type RuntimeEvent = RuntimeEvent;
    type RuntimeCall = RuntimeCall;
    type CallFilter = frame_support::traits::Nothing;
    type CallStack = [Frame<Self>; 31];
    type WeightPrice = Pink;
    type WeightInfo = weights::PinkWeights<Self>;
    type ChainExtension = extension::PinkExtension;
    type DeletionQueueDepth = DeletionQueueDepth;
    type DeletionWeightLimit = DeletionWeightLimit;
    type Schedule = DefaultSchedule;
    type DepositPerByte = DepositPerStorageByte;
    type DepositPerItem = DepositPerStorageItem;
    type AddressGenerator = Pink;
    type MaxCodeLen = MaxCodeLen;
    type MaxStorageKeyLen = MaxStorageKeyLen;
}

#[test]
fn detect_parameter_changes() {
    insta::assert_debug_snapshot!((
        <PinkRuntime as frame_system::Config>::BlockWeights::get(),
        <PinkRuntime as Config>::Schedule::get(),
        <PinkRuntime as Config>::DeletionQueueDepth::get(),
        <PinkRuntime as Config>::DeletionWeightLimit::get(),
        <PinkRuntime as Config>::MaxCodeLen::get(),
        <PinkRuntime as Config>::MaxStorageKeyLen::get(),
    ));
}

#[derive(Clone, Copy)]
pub enum CallMode {
    Query,
    Command,
}

pub trait EventCallbacks {
    fn emit_log(&self, contract: &AccountId, in_query: bool, level: u8, message: String);
}

pub type BoxedEventCallbacks = Box<dyn EventCallbacks>;

pub struct CallModeInfo {
    pub mode: CallMode,
    pub worker_pubkey: EcdhPublicKey,
}

struct CallInfo {
    mode: CallMode,
    start_at: Instant,
    callbacks: Option<BoxedEventCallbacks>,
}

environmental::environmental!(call_info: CallInfo);

static WORKER_PUBKEY: RwLock<EcdhPublicKey> = RwLock::new([0; 32]);

pub fn set_worker_pubkey(key: EcdhPublicKey) {
    *WORKER_PUBKEY.write().unwrap() = key;
}

pub fn using_mode<T>(
    mode: CallMode,
    callbacks: Option<BoxedEventCallbacks>,
    f: impl FnOnce() -> T,
) -> T {
    let mut info = CallInfo {
        mode,
        start_at: Instant::now(),
        callbacks,
    };
    call_info::using(&mut info, f)
}

pub fn get_call_mode_info() -> Option<CallModeInfo> {
    call_info::with(|info| CallModeInfo {
        mode: info.mode,
        worker_pubkey: *WORKER_PUBKEY.read().unwrap(),
    })
}

pub fn get_call_elapsed() -> Option<Duration> {
    call_info::with(|info| info.start_at.elapsed())
}

pub fn emit_log(id: &AccountId, level: u8, msg: String) {
    call_info::with(|info| {
        if let Some(callbacks) = &info.callbacks {
            callbacks.emit_log(id, matches!(info.mode, CallMode::Query), level, msg);
        }
    });
}

#[cfg(test)]
mod tests {
    #![allow(clippy::type_complexity)]
    use super::*;

    use frame_support::{assert_ok, traits::Currency};
    use pallet_contracts::Config;
    use sp_runtime::{traits::Hash, AccountId32};

    use crate::{
        runtime::{Contracts, PinkRuntime, RuntimeOrigin as Origin},
        storage::new_in_memory_backend,
        types::Balance,
        Contract, Storage, TransactionArguments,
    };

    pub use frame_support::weights::Weight;

    const ALICE: AccountId32 = AccountId32::new([1u8; 32]);
    const TREASURY: AccountId32 = AccountId32::new([2u8; 32]);
    const ENOUGH: Balance = Balance::MAX.saturating_div(32);
    const FLIPPER: &[u8] = include_bytes!("../tests/fixtures/flip/flip.wasm");
    const CENT: u128 = 10_000_000_000;

    pub fn compile_wat<T>(wat_bytes: &[u8]) -> wat::Result<(Vec<u8>, <T::Hashing as Hash>::Output)>
    where
        T: frame_system::Config,
    {
        let wasm_binary = wat::parse_bytes(wat_bytes)?.into_owned();
        let code_hash = T::Hashing::hash(&wasm_binary);
        Ok((wasm_binary, code_hash))
    }

    #[test]
    pub fn contract_test() {
        use scale::Encode;

        let (wasm, code_hash) =
            compile_wat::<PinkRuntime>(include_bytes!("../tests/fixtures/event_size.wat")).unwrap();

        exec::execute_with(|| {
            _ = Balances::deposit_creating(&ALICE, Balance::MAX.saturating_div(2));
            Contracts::instantiate_with_code(
                Origin::signed(ALICE),
                ENOUGH,
                Weight::MAX,
                None,
                wasm,
                vec![],
                vec![],
            )
            .unwrap();
            let addr = Contracts::contract_address(&ALICE, &code_hash, &[]);

            Contracts::call(
                Origin::signed(ALICE),
                addr,
                0,
                Weight::MAX,
                None,
                <PinkRuntime as Config>::Schedule::get()
                    .limits
                    .payload_len
                    .encode(),
            )
            .unwrap();
        });
        log::info!("contract OK");
    }

    #[test]
    pub fn crypto_hashes_test() {
        pub const ALICE: AccountId32 = AccountId32::new([1u8; 32]);
        pub const GAS_LIMIT: Weight =
            Weight::from_ref_time(100_000_000_000).set_proof_size(256u64 * 1024 * 1024);

        let (wasm, code_hash) =
            compile_wat::<PinkRuntime>(include_bytes!("../tests/fixtures/crypto_hashes.wat"))
                .unwrap();

        exec::execute_with(|| {
            _ = Balances::deposit_creating(&ALICE, Balance::MAX.saturating_div(2));
            // Instantiate the CRYPTO_HASHES contract.
            assert_ok!(Contracts::instantiate_with_code(
                Origin::signed(ALICE),
                1_000_000_000_000_000,
                GAS_LIMIT,
                None,
                wasm,
                vec![],
                vec![],
            ));
            let addr = Contracts::contract_address(&ALICE, &code_hash, &[]);
            // Perform the call.
            let input = b"_DEAD_BEEF";
            use sp_io::hashing::*;
            // Wraps a hash function into a more dynamic form usable for testing.
            macro_rules! dyn_hash_fn {
                ($name:ident) => {
                    Box::new(|input| $name(input).as_ref().to_vec().into_boxed_slice())
                };
            }
            // All hash functions and their associated output byte lengths.
            let test_cases: &[(Box<dyn Fn(&[u8]) -> Box<[u8]>>, usize)] = &[
                (dyn_hash_fn!(sha2_256), 32),
                (dyn_hash_fn!(keccak_256), 32),
                (dyn_hash_fn!(blake2_256), 32),
                (dyn_hash_fn!(blake2_128), 16),
            ];
            // Test the given hash functions for the input: "_DEAD_BEEF"
            for (n, (hash_fn, expected_size)) in test_cases.iter().enumerate() {
                // We offset data in the contract tables by 1.
                let mut params = vec![(n + 1) as u8];
                params.extend_from_slice(input);
                let result =
                    Contracts::bare_call(ALICE, addr.clone(), 0, GAS_LIMIT, None, params, false, pallet_contracts::Determinism::Deterministic)
                        .result
                        .unwrap();
                assert!(!result.did_revert());
                let expected = hash_fn(input.as_ref());
                assert_eq!(&result.data[..*expected_size], &*expected);
            }
        })
    }

    #[test]
    pub fn gas_limit_works() {
        let mut storage = crate::Storage::new(new_in_memory_backend());

        let gas_price = 2;
        let deposit_per_item = CENT * 8;
        let deposit_per_byte = CENT * 2;

        storage.setup(gas_price, deposit_per_item, deposit_per_byte, &TREASURY);

        let upload_result = storage.upload_code(&ALICE, FLIPPER.to_vec(), true);
        assert!(upload_result.is_err());
        let upload_result = storage.upload_sidevm_code(&ALICE, FLIPPER.to_vec());
        assert!(upload_result.is_err());

        let total_issue = Balance::MAX.saturating_div(2);

        storage.execute_mut(false, None, || {
            _ = Balances::deposit_creating(&ALICE, total_issue);
        });

        let upload_result = storage.upload_sidevm_code(&ALICE, FLIPPER.to_vec());
        assert_ok!(upload_result);

        let upload_result = storage.upload_code(&ALICE, FLIPPER.to_vec(), true);
        assert_ok!(&upload_result);
        let code_hash = upload_result.unwrap();

        // Flipper::default()
        let default_selector = 0xed4b9d1b_u32.to_be_bytes().to_vec();
        let result = Contract::new(
            code_hash,
            default_selector,
            vec![],
            false,
            tx_args(&mut storage),
        );
        assert_ok!(&result);

        let flipper = result.unwrap().0;

        let mut prev_free_balance = storage.free_balance(&ALICE);

        // The contract flipper instantiated
        let fn_flip = 0x633aa551_u32.to_be_bytes().to_vec();
        let fn_get = 0x2f865bd9_u32.to_be_bytes();

        let init_value: bool = {
            let mut args = tx_args(&mut storage);
            args.gas_free = true;
            flipper
                .call_with_selector(fn_get, (), true, args)
                .0
                .unwrap()
        };

        // Estimate gas
        let est_result = {
            let mut args = tx_args(&mut storage);
            args.gas_free = true;
            let result = flipper.bare_call(fn_flip.clone(), true, args).0;
            assert_ok!(&result.result);
            assert_eq!(storage.free_balance(&ALICE), prev_free_balance);
            result
        };

        {
            // Zero gas limit should not work
            let gas_limit = 0;
            let mut args = tx_args(&mut storage);
            args.gas_free = false;
            args.gas_limit = Weight::from_ref_time(gas_limit);
            let result = flipper.bare_call(fn_flip.clone(), false, args).0;
            assert!(result.result.is_err());
            assert_eq!(prev_free_balance, storage.free_balance(&ALICE));

            // Should NOT flipped
            let value: bool = {
                let mut args = tx_args(&mut storage);
                args.gas_free = true;
                flipper
                    .call_with_selector(fn_get, (), true, args)
                    .0
                    .unwrap()
            };
            assert_eq!(init_value, value);
        }

        {
            let gas_limit = est_result.gas_required / 2;
            let mut args = tx_args(&mut storage);
            args.gas_free = false;
            args.gas_limit = gas_limit;
            let result = flipper.bare_call(fn_flip.clone(), false, args).0;
            assert!(result.result.is_err());
            assert_eq!(
                prev_free_balance - result.gas_consumed.ref_time() as u128 * gas_price,
                storage.free_balance(&ALICE)
            );

            prev_free_balance = storage.free_balance(&ALICE);

            // Should NOT flipped
            let value: bool = {
                let mut args = tx_args(&mut storage);
                args.gas_free = true;
                flipper
                    .call_with_selector(fn_get, (), true, args)
                    .0
                    .unwrap()
            };
            assert_eq!(init_value, value);
        }

        {
            let gas_limit = est_result.gas_required;
            let mut args = tx_args(&mut storage);
            args.gas_free = false;
            args.gas_limit = gas_limit;
            let result = flipper.bare_call(fn_flip, false, args).0;
            assert_ok!(result.result);
            let cost = prev_free_balance - storage.free_balance(&ALICE);
            assert_eq!(cost, result.gas_consumed.ref_time() as u128 * gas_price);

            // Should flipped
            let value: bool = {
                let mut args = tx_args(&mut storage);
                args.gas_free = true;
                flipper
                    .call_with_selector(fn_get, (), true, args)
                    .0
                    .unwrap()
            };
            assert_eq!(!init_value, value);
        }
    }

    fn tx_args(storage: &mut Storage) -> TransactionArguments {
        TransactionArguments {
            origin: ALICE.clone(),
            now: 1,
            block_number: 1,
            storage,
            transfer: 0,
            gas_limit: Weight::MAX,
            gas_free: false,
            storage_deposit_limit: None,
            callbacks: None,
        }
    }

    pub mod exec {
        use sp_runtime::traits::BlakeTwo256;
        use sp_state_machine::{
            backend::AsTrieBackend, Ext, OverlayedChanges, StorageTransactionCache,
        };

        pub type InMemoryBackend = sp_state_machine::InMemoryBackend<BlakeTwo256>;

        pub fn execute_with<R>(f: impl FnOnce() -> R) -> R {
            let state = InMemoryBackend::default();
            let backend = state.as_trie_backend();

            let mut overlay = OverlayedChanges::default();
            overlay.start_transaction();
            let mut cache = StorageTransactionCache::default();
            let mut ext = Ext::new(&mut overlay, &mut cache, backend, None);
            let r = sp_externalities::set_and_run_with_externalities(&mut ext, f);
            overlay
                .commit_transaction()
                .expect("BUG: mis-paired transaction");
            r
        }
    }
}
