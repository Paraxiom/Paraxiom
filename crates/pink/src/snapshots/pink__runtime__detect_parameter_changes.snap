---
source: crates/pink/src/runtime.rs
assertion_line: 151
expression: "(<PinkRuntime as frame_system::Config>::BlockWeights::get(),\n    <PinkRuntime as Config>::Schedule::get(),\n    <PinkRuntime as Config>::DeletionQueueDepth::get(),\n    <PinkRuntime as Config>::DeletionWeightLimit::get(),\n    <PinkRuntime as Config>::MaxCodeLen::get(),\n    <PinkRuntime as Config>::MaxStorageKeyLen::get())"
---
(
    BlockWeights {
        base_block: Weight {
            ref_time: 358523000,
            proof_size: 0,
        },
        max_block: Weight {
            ref_time: 2000000000000,
            proof_size: 18446744073709551615,
        },
        per_class: PerDispatchClass {
            normal: WeightsPerClass {
                base_extrinsic: Weight {
                    ref_time: 98974000,
                    proof_size: 0,
                },
                max_extrinsic: Some(
                    Weight {
                        ref_time: 1299901026000,
                        proof_size: 11990383647911208550,
                    },
                ),
                max_total: Some(
                    Weight {
                        ref_time: 1500000000000,
                        proof_size: 13835058055282163711,
                    },
                ),
                reserved: Some(
                    Weight {
                        ref_time: 0,
                        proof_size: 0,
                    },
                ),
            },
            operational: WeightsPerClass {
                base_extrinsic: Weight {
                    ref_time: 98974000,
                    proof_size: 0,
                },
                max_extrinsic: Some(
                    Weight {
                        ref_time: 1799901026000,
                        proof_size: 16602069666338596454,
                    },
                ),
                max_total: Some(
                    Weight {
                        ref_time: 2000000000000,
                        proof_size: 18446744073709551615,
                    },
                ),
                reserved: Some(
                    Weight {
                        ref_time: 500000000000,
                        proof_size: 4611686018427387904,
                    },
                ),
            },
            mandatory: WeightsPerClass {
                base_extrinsic: Weight {
                    ref_time: 98974000,
                    proof_size: 0,
                },
                max_extrinsic: None,
                max_total: None,
                reserved: None,
            },
        },
    },
    Schedule {
        limits: Limits {
            event_topics: 4,
            stack_height: None,
            globals: 256,
            parameters: 128,
            memory_pages: 64,
            table_size: 4096,
            br_table_size: 256,
            subject_len: 32,
            call_depth: 32,
            payload_len: 16384,
        },
        instruction_weights: InstructionWeights {
            version: "3 ps",
            fallback: "8.0 ns",
            i64const: "4.7 ns",
            i64load: "20.4 ns",
            i64store: "17.5 ns",
            select: "5.1 ns",
            r#if: "12.3 ns",
            br: "7.1 ns",
            br_if: "7.4 ns",
            br_table: "9.4 ns",
            br_table_per_entry: "30 ps",
            call: "64.0 ns",
            call_indirect: "78.1 ns",
            call_indirect_per_param: "795 ps",
            local_get: "5.9 ns",
            local_set: "5.3 ns",
            local_tee: "5.3 ns",
            global_get: "10.3 ns",
            global_set: "10.8 ns",
            memory_current: "4.7 ns",
            memory_grow: "11.4 µs",
            i64clz: "4.2 ns",
            i64ctz: "4.3 ns",
            i64popcnt: "4.2 ns",
            i64eqz: "4.2 ns",
            i64extendsi32: "4.2 ns",
            i64extendui32: "4.2 ns",
            i32wrapi64: "4.2 ns",
            i64eq: "4.2 ns",
            i64ne: "4.2 ns",
            i64lts: "4.3 ns",
            i64ltu: "4.2 ns",
            i64gts: "4.3 ns",
            i64gtu: "4.3 ns",
            i64les: "4.3 ns",
            i64leu: "4.3 ns",
            i64ges: "4.3 ns",
            i64geu: "4.3 ns",
            i64add: "4.3 ns",
            i64sub: "4.3 ns",
            i64mul: "4.3 ns",
            i64divs: "11.6 ns",
            i64divu: "8.9 ns",
            i64rems: "11.5 ns",
            i64remu: "9.4 ns",
            i64and: "4.4 ns",
            i64or: "4.4 ns",
            i64xor: "4.4 ns",
            i64shl: "4.4 ns",
            i64shrs: "4.5 ns",
            i64shru: "4.5 ns",
            i64rotl: "4.8 ns",
            i64rotr: "4.6 ns",
        },
        host_fn_weights: HostFnWeights {
            caller: "442.7 ns",
            is_contract: "2.6 µs",
            code_hash: "3.4 µs",
            own_code_hash: "485.0 ns",
            caller_is_origin: "197.8 ns",
            address: "442.5 ns",
            gas_left: "439.3 ns",
            balance: "1.4 µs",
            value_transferred: "440.2 ns",
            minimum_balance: "438.8 ns",
            block_number: "432.6 ns",
            now: "434.8 ns",
            weight_to_fee: "1.2 µs",
            gas: "195.4 ns",
            input: "418.5 ns",
            input_per_byte: "117 ps",
            r#return: "2.1 µs",
            return_per_byte: "179 ps",
            terminate: "52.9 µs",
            random: "1.7 µs",
            deposit_event: "3.0 µs",
            deposit_event_per_topic: "2.2 µs",
            deposit_event_per_byte: "822 ps",
            debug_message: "331.7 ns",
            set_storage: "5.2 µs",
            set_storage_per_new_byte: "1.1 ns",
            set_storage_per_old_byte: "839 ps",
            set_code_hash: "13.7 µs",
            clear_storage: "5.1 µs",
            clear_storage_per_byte: "863 ps",
            contains_storage: "3.8 µs",
            contains_storage_per_byte: "756 ps",
            get_storage: "4.1 µs",
            get_storage_per_byte: "1.9 ns",
            take_storage: "5.4 µs",
            take_storage_per_byte: "2.0 ns",
            transfer: "17.1 µs",
            call: "218.8 µs",
            delegate_call: "217.1 µs",
            call_transfer_surcharge: "14.5 µs",
            call_per_cloned_byte: "121.6 ns",
            instantiate: "285.4 µs",
            instantiate_transfer_surcharge: "0 ps",
            instantiate_per_salt_byte: "1.5 ns",
            hash_sha2_256: "688.3 ns",
            hash_sha2_256_per_byte: "3.9 ns",
            hash_keccak_256: "845.3 ns",
            hash_keccak_256_per_byte: "3.0 ns",
            hash_blake2_256: "564.6 ns",
            hash_blake2_256_per_byte: "1.2 ns",
            hash_blake2_128: "560.6 ns",
            hash_blake2_128_per_byte: "1.2 ns",
            ecdsa_recover: "37.1 µs",
            ecdsa_to_eth_address: "25.8 µs",
        },
    },
    1024,
    Weight {
        ref_time: 500000000000,
        proof_size: 0,
    },
    2097152,
    128,
)
